sympy/combinatorics/fp_groups.py

```diff
@@ -279,6 +285,54 @@ def __str__(self):
     __repr__ = __str__
 
 
+class FpSubgroup(DefaultPrinting):
+    '''
+    The class implementing a subgroup of an FpGroup or a FreeGroup
+    (only finite index subgroups are supported at this point). This
+    is to be used if one wishes to check if an element of the original
+    group belongs to the subgroup
+
+    '''
+    def __init__(self, G, gens):
+        super(FpSubgroup,self).__init__()
+        self.parent = G
+        self.generators = list(set([g for g in gens if g != G.identity]))
+        self.C = None
+
+    def __contains__(self, g):
+        gens = [s for s in g.contains_generators()]
+        if any([s not in self.generators for s in gens]):
+            return False
+        elif isinstance(self.parent, FreeGroup):
+            return True
+        if self.C is None:
+            C = self.parent.coset_enumeration(self.generators)
+            self.C = C
+        i = 0
+        C = self.C
+        for j in range(len(g)):
+            i = C.table[i][C.A_dict[g[j]]]
+        return i == 0
+
+    def order(self):
+        # This is valid because `len(self.C.table)` (the index of the subgroup)
+        # will always be finite - otherwise coset enumeration doesn't terminate
```

Comment:
The index of a subgroup need not be finite, but a subgroup of a free group is also a free group, hence infinite or trivial.